File: android/sdk/src/main/kotlin/com/ary/push/model/PushPermissionStatus.kt
package com.ary.push.model

/**
 * Notification permission state, normalised across Android and iOS.
 *
 * Android only ever reports [NOT_DETERMINED], [GRANTED] or [DENIED]. [PROVISIONAL],
 * [EPHEMERAL] and [RESTRICTED] exist so that the same model can describe iOS, keeping host
 * application code portable.
 */
public enum class PushPermissionStatus {

    /** The user has not been asked yet. */
    NOT_DETERMINED,

    /** Notifications may be posted. */
    GRANTED,

    /** The user declined, or notifications are switched off for the application. */
    DENIED,

    /** iOS only: quiet delivery granted without an explicit prompt. */
    PROVISIONAL,

    /** iOS only: temporary authorization granted to an App Clip. */
    EPHEMERAL,

    /** iOS only: notifications are restricted by policy and cannot be requested. */
    RESTRICTED;

    /** True when the SDK is allowed to post notifications, quietly or otherwise. */
    public val isAuthorized: Boolean
        get() = this == GRANTED || this == PROVISIONAL || this == EPHEMERAL
}

File: ios/Sources/ARYPush/Model/PushPermissionStatus.swift
import Foundation
import UserNotifications

/// Notification permission state, normalised across iOS and Android.
///
/// `UNAuthorizationStatus` has states Android does not (`provisional`, `ephemeral`) and Android
/// has behaviour iOS does not, so both are mapped onto this one model. Host application code
/// then reads the same values on every platform.
public enum PushPermissionStatus: String, Equatable {

    /// The user has not been asked yet.
    case notDetermined

    /// Notifications may be presented.
    case granted

    /// The user declined, or notifications are switched off for the application.
    case denied

    /// Quiet delivery granted without an explicit prompt.
    case provisional

    /// Temporary authorization granted to an App Clip.
    case ephemeral

    /// Restricted by policy; a prompt is not possible.
    case restricted

    /// True when the SDK is allowed to present notifications, quietly or otherwise.
    public var isAuthorized: Bool {
        self == .granted || self == .provisional || self == .ephemeral
    }

    /// Maps the platform value onto the shared model.
    static func from(_ status: UNAuthorizationStatus) -> PushPermissionStatus {
        switch status {
        case .notDetermined: return .notDetermined
        case .denied: return .denied
        case .authorized: return .granted
        case .provisional: return .provisional
        case .ephemeral: return .ephemeral
        @unknown default:
            // A future iOS state the SDK has not seen. Reporting `denied` is the safe answer:
            // it never claims a capability the SDK might not have.
            return .denied
        }
    }
}
